### PR TITLE
MaxContentLength support (streaming)

### PR DESCRIPTION
**Pull request by kciesielski**
Started as a solution to https://github.com/softwaremill/tapir/issues/3056 for Netty, but addresses more backends in general.
This PR introduces a possibility to add a `MaxContentLength` attribute to any endpoint. When set, decoding `RequestBody` in supported backends prevents loading too much into memory, and fails if the limit is exceeded.
As a result, `DefaultExceptionHandler` returns `HTTP 413 Payload Too Large`.

*Streaming only*
This is only a part of the full feature. The PR covers streaming support for streaming request body, excluding:
- Armeria streams
- Vertx streams
- zio1 streams

In a follow-up PR, I'll add:
- Support for non-streaming body (direct byte arrays, files) 
- Documentation
- API extension to allow writing `endpoint.maxContentLength(x)` instead of setting an attribute.

### CODE REVIEW

**Comment by kciesielski** [comment]
@mergifyio update

File: server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/internal/NettyCatsRequestBody.scala [hunk 1]
```diff
@@ -48,17 +48,18 @@ private[netty] class NettyCatsRequestBody[F[_]](createFile: ServerRequest => F[T
...
       .eval(StreamSubscriber[F, HttpContent](NettyRequestBody.DefaultChunkSize))
       .flatMap(s => s.sub.stream(Sync[F].delay(nettyRequest.subscribe(s))))
       .flatMap(httpContent => fs2.Stream.chunk(Chunk.byteBuffer(httpContent.content.nioBuffer())))
+    maxBytes.map(Fs2Streams.limitBytes(stream, _)).getOrElse(stream)
   }
 
   private def nettyRequestBytes(serverRequest: ServerRequest): F[Array[Byte]] = serverRequest.underlying match {
     case req: FullHttpRequest   => monad.delay(ByteBufUtil.getBytes(req.content()))
-    case _: StreamedHttpRequest => toStream(serverRequest).compile.to(Chunk).map(_.toArray[Byte])
+    case _: StreamedHttpRequest => toStream(serverRequest, maxBytes = None).compile.to(Chunk).map(_.toArray[Byte]) // TODO
```
Comment (adamw) [inline server/netty-server/cats/src/main/scala/sttp/tapir/server/netty/internal/NettyCatsRequestBody.scala:62]: TODO ?:)

**Comment by adamw** [comment]
It seems there's a half-done maxContentLegth implementation for netty, in `NettyServerHandler`, plus a test in `ServerBasicTests`, which is only enabled for netty, though. Would be good to clean this up in the next PR as well :)

File: server/netty-server/zio/src/main/scala/sttp/tapir/server/netty/internal/NettyZioRequestBody.scala [hunk 2]
```diff
@@ -44,17 +44,17 @@ private[netty] class NettyZioRequestBody[Env](createFile: ServerRequest => RIO[E
...
       .asInstanceOf[StreamedHttpRequest]
       .toZIOStream()
       .flatMap(httpContent => ZStream.fromChunk(Chunk.fromByteBuffer(httpContent.content.nioBuffer())))
+    maxBytes.map(ZioStreams.limitBytes(stream, _)).getOrElse(stream)
   }
 
   private def nettyRequestBytes(serverRequest: ServerRequest): RIO[Env, Array[Byte]] = serverRequest.underlying match {
     case req: FullHttpRequest   => ZIO.succeed(ByteBufUtil.getBytes(req.content()))
-    case _: StreamedHttpRequest => toStream(serverRequest).run(ZSink.collectAll[Byte]).map(_.toArray)
+    case _: StreamedHttpRequest => toStream(serverRequest, maxBytes = None).run(ZSink.collectAll[Byte]).map(_.toArray) // TODO
```
Comment (adamw) [inline server/netty-server/zio/src/main/scala/sttp/tapir/server/netty/internal/NettyZioRequestBody.scala:57]: that's for the next PRs?

Comment (kciesielski) [reply server/netty-server/zio/src/main/scala/sttp/tapir/server/netty/internal/NettyZioRequestBody.scala:57]: Yes, coming soon :) https://github.com/softwaremill/tapir/pull/3337

**Comment by kciesielski** [comment]
> It seems there's a half-done maxContentLegth implementation for netty, in `NettyServerHandler`, plus a test in `ServerBasicTests`, which is only enabled for netty, though. Would be good to clean this up in the next PR as well :)

That current netty-specific maxContentLength support is actually an interesting problem to discuss. It's 
1) global, not per endpoint
2) affects both responses and requests
3) does not work for streaming

If we  remove it entirely in favor of the new per-endpoint solution, we will lose features of 1) global scope 2) limiting response body.
If we keep it, it may be confusing to understand why there are 2 ways of setting max content length for Netty. Maybe we should limit it to responses only? Or deprecate it and observe who complains that they actually wanted 1 and 2, and not a per-endpoint request limiting?

**Comment by adamw** [comment]
Ideally, we want both: a global setting in `ServerOptions`, with a per-endpoint override (including disabling the global value). Though you can always map over all server endpoints and add the attribute, so maybe that's not needed 🤔 

But we should definitely have one solution, so let's keep the "new" one only. I can't see limiting response size as a useful feature (you can do it in the business logic anyway)

**Pull request merged**